File: python/context_consent.py
"""Context collection consent (Full / Minimal / Off) for Python connectors.

Resolution order:
  1. env ``PATCHERLY_CONTEXT_CONSENT``
  2. file ``{PATCHERLY_CACHE_DIR}/context_consent`` (default cache ``.patcherly_cache``)
  3. default ``full``
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Literal, Tuple

ConsentTier = Literal["full", "minimal", "off"]
VALID_TIERS = frozenset({"full", "minimal", "off"})
DEFAULT_TIER: ConsentTier = "full"


def cache_dir() -> Path:
    raw = (os.getenv("PATCHERLY_CACHE_DIR") or "").strip() or ".patcherly_cache"
    path = Path(raw)
    path.mkdir(parents=True, exist_ok=True)
    return path


def consent_file_path() -> Path:
    return cache_dir() / "context_consent"


def _normalize(raw: str | None) -> ConsentTier | None:
    if raw is None:
        return None
    v = raw.strip().lower()
    if v in VALID_TIERS:
        return v  # type: ignore[return-value]
    return None


def get_context_consent() -> Tuple[ConsentTier, str]:
    """Return (tier, source) where source is env|file|default."""
    env_raw = os.getenv("PATCHERLY_CONTEXT_CONSENT")
    env_tier = _normalize(env_raw)
    if env_tier is not None:
        return env_tier, "env"
    if env_raw is not None and str(env_raw).strip():
        # Invalid env — fall through to file/default but prefer not crashing
        pass
    try:
        path = consent_file_path()
        if path.is_file():
            file_tier = _normalize(path.read_text(encoding="utf-8"))
            if file_tier is not None:
                return file_tier, "file"
    except OSError:
        pass
    return DEFAULT_TIER, "default"


def set_context_consent(tier: str) -> ConsentTier:
    normalized = _normalize(tier)
    if normalized is None:
        raise ValueError(f"Invalid consent tier {tier!r}; expected full|minimal|off")
    path = consent_file_path()
    path.write_text(normalized + "\n", encoding="utf-8")
    return normalized

File: nodejs/context_consent.js
'use strict';
/**
 * Context collection consent (Full / Minimal / Off) for Node.js connectors.
 *
 * Resolution order:
 *   1. env  PATCHERLY_CONTEXT_CONSENT
 *   2. file {PATCHERLY_CACHE_DIR}/context_consent  (default cache .patcherly_cache)
 *   3. default: full
 *
 * Mirrors connectors/python/context_consent.py exactly.
 */

const fs = require('fs');
const path = require('path');

const VALID_TIERS = new Set(['full', 'minimal', 'off']);
const DEFAULT_TIER = 'full';

/**
 * Resolve the cache directory (create if missing) and return its path.
 * @returns {string}
 */
function cacheDir() {
  const raw = (process.env.PATCHERLY_CACHE_DIR || '').trim();
  const dir = raw !== '' ? raw : '.patcherly_cache';
  try {
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
  } catch (_) { /* best-effort */ }
  return dir;
}

/**
 * Full path to the consent tier file.
 * @returns {string}
 */
function consentFilePath() {
  return path.join(cacheDir(), 'context_consent');
}

/**
 * Normalize a raw string to a valid tier or return null for invalid values.
 * @param {string|null|undefined} raw
 * @returns {string|null}
 */
function _normalize(raw) {
  if (raw == null) return null;
  const v = String(raw).trim().toLowerCase();
  return VALID_TIERS.has(v) ? v : null;
}

/**
 * Return { tier, source } where source is 'env'|'file'|'default'.
 *
 * Invalid values in env or file are treated as absent (fall through to next),
 * matching the Python behavior exactly.
 *
 * @returns {{ tier: string, source: string }}
 */
function getContextConsent() {
  const envRaw = process.env.PATCHERLY_CONTEXT_CONSENT;
  if (envRaw !== undefined) {
    const tier = _normalize(envRaw);
    if (tier !== null) return { tier, source: 'env' };
    // Invalid env value — fall through (mirrors Python behavior).
  }
  try {
    const filePath = consentFilePath();
    if (fs.existsSync(filePath)) {
      const tier = _normalize(fs.readFileSync(filePath, 'utf8'));
      if (tier !== null) return { tier, source: 'file' };
    }
  } catch (_) { /* IO error — fall through to default */ }
  return { tier: DEFAULT_TIER, source: 'default' };
}

/**
 * Write the normalized tier to the consent file and return it.
 *
 * @param {string} tier
 * @returns {string}  the normalized tier
 * @throws {Error} for unrecognized tier values
 */
function setContextConsent(tier) {
  const normalized = _normalize(tier);
  if (normalized === null) {
    throw new Error(`Invalid consent tier ${JSON.stringify(tier)}; expected full|minimal|off`);
  }
  fs.writeFileSync(consentFilePath(), normalized + '\n', 'utf8');
  return normalized;
}

module.exports = { getContextConsent, setContextConsent, consentFilePath };
